File: x86_trace_generator/pin/bindings/glue/pinglueserver/tests/Scripts/malloctrace_probe.py
#
# Copyright (C) 2024-2024 Intel Corporation.
# SPDX-License-Identifier: MIT
#

import pin

mallocCount = 0

output_file = open("script_malloctrace_probe.out", "w")

def BeforeExit():
    output_file.close()

def Arg1Before(name, size):
    size = hex(size) if size else 0
    output_file.write(f"{name}({size})\n")

def MallocAfter(ret):
    global mallocCount
    mallocCount += 1
    output_file.write(f"  returns {hex(ret)} [{mallocCount}]\n")

def image_instrumentation_cb(img):
    mallocRtn = pin.RTN_FindByName(img, "malloc")
    if(pin.RTN_Valid(mallocRtn)):
        assert(pin.RTN_IsSafeForProbedInsertion(mallocRtn))
        pin.RTN_InsertCallProbed(mallocRtn, pin.IPOINT_BEFORE, Arg1Before, pin.IARG_PYOBJ, "malloc", pin.IARG_FUNCARG_ENTRYPOINT_VALUE, 0)
        
        proto_malloc_after = pin.PROTO_Allocate(pin.PIN_PARG_POINTER, pin.CALLINGSTD_DEFAULT, "malloc", pin.PIN_PARG_SIZE_T)
        pin.RTN_InsertCallProbed(mallocRtn, pin.IPOINT_AFTER, MallocAfter, pin.IARG_PROTOTYPE, proto_malloc_after, pin.IARG_FUNCRET_EXITPOINT_VALUE)
        pin.PROTO_Free(proto_malloc_after)
	
    freeRtn = pin.RTN_FindByName(img, "free")
    if (pin.RTN_Valid(freeRtn)):
        assert(pin.RTN_IsSafeForProbedInsertion(freeRtn))
        pin.RTN_InsertCallProbed(freeRtn, pin.IPOINT_BEFORE, Arg1Before, pin.IARG_PYOBJ, "free", pin.IARG_FUNCARG_ENTRYPOINT_VALUE, 0)
    
    exitRtn = pin.RTN_FindByName(img, "_exit")
    if (pin.RTN_Valid(exitRtn)):
        pin.RTN_InsertCallProbed(exitRtn, pin.IPOINT_BEFORE, BeforeExit)

pin.IMG_AddInstrumentFunction(image_instrumentation_cb)

File: x86_trace_generator/pin/bindings/glue/pinglueserver/tests/NativeTools/malloctrace_probe.cpp
/*
 * Copyright (C) 2024-2024 Intel Corporation.
 * SPDX-License-Identifier: MIT
 */

#include "pin.H"
#include <iostream>
#include <fstream>
using std::cerr;
using std::endl;
using std::hex;
using std::ios;
using std::string;

/* ===================================================================== */
/* Names of malloc and free */
/* ===================================================================== */
#define MALLOC "malloc"
#define FREE "free"

/* ===================================================================== */
/* Global Variables */
/* ===================================================================== */

std::ofstream TraceFile;

/* ===================================================================== */
/* Commandline Switches */
/* ===================================================================== */

KNOB< string > KnobOutputFile(KNOB_MODE_WRITEONCE, "pintool", "o", "malloctrace_probe.out", "specify trace file name");

/* ===================================================================== */

/* ===================================================================== */
/* Analysis routines                                                     */
/* ===================================================================== */

VOID Arg1Before(CHAR* name, ADDRINT size) { TraceFile << name << "(" << size << ")" << endl; }

VOID MallocAfter(ADDRINT ret) { TraceFile << "  returns " << ret << endl; }

/* ===================================================================== */
/* Instrumentation routines                                              */
/* ===================================================================== */

VOID Image(IMG img, VOID* v)
{
    // Instrument the malloc() and free() functions.  Print the input argument
    // of each malloc() or free(), and the return value of malloc().
    //
    //  Find the malloc() function.
    RTN mallocRtn = RTN_FindByName(img, MALLOC);
    if (RTN_Valid(mallocRtn))
    {
        // Instrument malloc() to print the input argument value and the return value.
        RTN_InsertCallProbed(mallocRtn, IPOINT_BEFORE, (AFUNPTR)Arg1Before, IARG_ADDRINT, MALLOC, IARG_FUNCARG_ENTRYPOINT_VALUE,
                             0, IARG_END);

        PROTO proto_malloc_after = PROTO_Allocate(PIN_PARG(void*), CALLINGSTD_DEFAULT, MALLOC, PIN_PARG(size_t), PIN_PARG_END());
        if (RTN_InsertCallProbed(mallocRtn, IPOINT_AFTER, (AFUNPTR)MallocAfter, IARG_PROTOTYPE, proto_malloc_after,
                                 IARG_FUNCRET_EXITPOINT_VALUE, IARG_END))
        {
            ASSERTX(RTN_IsSafeForProbedInsertion(mallocRtn));
            std::cout << "Inserting calls before/after malloc in " << IMG_Name(img) << std::endl;
        }
        else
        {
            std::cout << "Cannot insert calls around " << RTN_Name(mallocRtn) << "() in " << IMG_Name(img) << std::endl;
            exit(1);
        }

        PROTO_Free(proto_malloc_after);
    }

    // Find the free() function.
    RTN freeRtn = RTN_FindByName(img, FREE);
    if (RTN_Valid(freeRtn))
    {
        ASSERTX(RTN_IsSafeForProbedInsertion(freeRtn));
        // Instrument free() to print the input argument value.
        RTN_InsertCallProbed(freeRtn, IPOINT_BEFORE, (AFUNPTR)Arg1Before, IARG_ADDRINT, FREE, IARG_FUNCARG_ENTRYPOINT_VALUE, 0,
                             IARG_END);
    }
}

/* ===================================================================== */
/* Print Help Message                                                    */
/* ===================================================================== */

INT32 Usage()
{
    cerr << "This tool produces a trace of calls to malloc." << endl;
    cerr << endl << KNOB_BASE::StringKnobSummary() << endl;
    return -1;
}

/* ===================================================================== */
/* Main                                                                  */
/* ===================================================================== */

int main(int argc, char* argv[])
{
    // Initialize pin & symbol manager
    PIN_InitSymbolsAlt(SYMBOL_INFO_MODE(UINT32(IFUNC_SYMBOLS) | UINT32(DEBUG_OR_EXPORT_SYMBOLS)));
    if (PIN_Init(argc, argv))
    {
        return Usage();
    }

    // Write to a file since cout and cerr maybe closed by the application
    TraceFile.open(KnobOutputFile.Value().c_str());
    TraceFile << hex;
    TraceFile.setf(ios::showbase);

    // Register Image to be called to instrument functions.
    IMG_AddInstrumentFunction(Image, 0);

    // Never returns
    PIN_StartProgramProbed();

    return 0;
}

/* ===================================================================== */
/* eof */
/* ===================================================================== */
